demos/instr.h

```diff
@@ -24,3 +30,60 @@ void CLFlush(const void *memory);
 
 // Measures the latency of memory read from a given address
 uint64_t ReadLatency(const void *memory);
+
+#ifdef __GNUC__
+// Inlines the afterspeculation label.
+__attribute__((always_inline))
+inline void afterspeculate() {
+  asm volatile(
+      "_afterspeculation:\n" // For MacOS.
+      "afterspeculation:\n"); // For Linux.
+}
+
+#if defined(__i386__) || defined(__x86_64__) || defined(_M_X64) || \
+    defined(_M_IX86)
+// Unwinds the stack until the given pointer, flushes the stack pointer and
+// returns.
+void UnwindStackAndSlowlyReturnTo(const void *address);
+#elif defined(__aarch64__)
+// Unwinds the stack until stored return address marked by a magic value,
+// flushes the stack pointer, loads the returns address to x30 and returns.
+void UnwindStackAndSlowlyReturn();
+
+// Push callee-saved registers and return address on stack and mark it with
+// magic value.
+__attribute__((always_inline))
+inline void BackupCalleeSavedRegsAndReturnAddress() {
+  asm volatile(
+      // Store the callee-saved regs.
+      "stp x19, x20, [sp, #-16]!\n"
+      "stp x21, x22, [sp, #-16]!\n"
+      "stp x23, x24, [sp, #-16]!\n"
+      "stp x25, x26, [sp, #-16]!\n"
+      "stp x27, x28, [sp, #-16]!\n"
+      "str x29, [sp, #-16]!\n"
+      // Push the afterspeculation address. We cannot pass it by function
+      // parameter as on x86/x64 because of relative/absolute address
+      // mismatches on ARM.
```

Comment:
This seems surprising. I can pass a function pointer on ARM and then call it in the function that receives it. What makes this different?

Reply:
The global vs. local relocation. When the variable is declared and extern and specified in assembler, it is linked as local, but accessed as global.

Reply:
I'm going to create a local handler for ARM.